react/src/components/SourceCodeViewer.tsx

```diff
@@ -0,0 +1,112 @@
+  children: string;
+  language: string;
+  wordWrap?: boolean;
+}
+
+const SourceCodeViewer: React.FC<SourceCodeViewerProps> = ({
+  children,
+  language,
+  wordWrap,
+}) => {
+  const timerRef = useRef<number>();
+  const { token } = theme.useToken();
+  const { isDarkMode } = useThemeMode();
+  const [isCopied, setIsCopied] = useState(false);
+  const [isMouseEntered, setIsMouseEntered] = useState(false);
+  const HighLightedCode = (
+    <Highlight
+      theme={isDarkMode ? themes.jettwaveDark : themes.jettwaveLight}
+      code={children || ''}
+      language={language || ''}
+    >
+      {({ className, style, tokens, getLineProps, getTokenProps }) => (
+        <pre
+          style={{
+            ...style,
+            width: '100%',
+            margin: 0,
+            padding: token.padding,
+            borderRadius: token.borderRadius,
+            overflow: wordWrap ? undefined : 'overlay',
+          }}
+        >
+          <code
+            style={{
+              whiteSpace: wordWrap ? 'pre-wrap' : undefined,
+              overflowWrap: wordWrap ? 'anywhere' : undefined,
+            }}
+          >
+            {_.map(tokens, (line, i) => {
+              return (
+                <span key={i} {...getLineProps({ line })}>
+                  {_.map(line, (token, key) => {
+                    /** The codes below fix some line wrapping errors during the conversion process. **/
+                    if (!token.content && key === line.length - 1) {
+                      return '\n';
+                    }
+                    if (line.length === 1 && token.content === ' ') {
+                      return '\n';
+                    }
+                    /** **/
+                    return (
+                      <span key={key} {...getTokenProps({ token })}>
+                        {token.content}
+                      </span>
+                    );
+                  })}
+                </span>
+              );
+            })}
+          </code>
+        </pre>
+      )}
+    </Highlight>
+  );
+
+  return (
+    <Flex
+      onMouseEnter={() => setIsMouseEntered(true)}
+      onMouseLeave={() => setIsMouseEntered(false)}
+    >
+      <CopyToClipboard
+        text={children}
+        onCopy={() => {
+          setIsCopied(true);
+          clearTimeout(timerRef.current);
+          timerRef.current = window.setTimeout(() => {
+            setIsCopied(false);
+          }, 2000);
+        }}
+      >
+        <Button
+          style={{
+            top: token.sizeSM,
+            right: token.sizeSM,
+            zIndex: token.zIndexBase + 1,
+            opacity: isMouseEntered ? 1 : 0,
+            backgroundColor: token.colorFillSecondary,
+            cursor: 'pointer',
+            position: 'absolute',
+            transition: 'opacity 0.2s ease-in-out',
+          }}
+          size="small"
+          icon={isCopied ? <CheckOutlined /> : <CopyOutlined />}
+        ></Button>
+      </CopyToClipboard>
```

Comment:
Lets' provide tooltip using Tooltip component to make it same ux of [copyable feature]([url](https://ant.design/components/typography#components-typography-demo-copyable)) of `Typhography` component.

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/XqC2uNFuj0wg8I60sMUh/a1da62c0-1a35-4baf-8ef5-70925dbcdc81.png)

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/XqC2uNFuj0wg8I60sMUh/3a63b977-efb1-4197-9cb8-e4bdd2f4069a.png)

Reply:
Applied